src/EthernetConnectionManager.h

```diff
@@ -0,0 +1,187 @@
+      break;
+    case CONNECTION_STATE_CONNECTED:
+      newInterval = CHECK_INTERVAL_CONNECTED;
+      break;
+    case CONNECTION_STATE_DISCONNECTED:
+      newInterval = CHECK_INTERVAL_DISCONNECTED;
+
+      break;
+  }
+  connectionTickTimeInterval = newInterval;
+  lastConnectionTickTime = millis();
+}
+
+void EthConnectionManager::check() {
+  char msgBuffer[120];
+  unsigned long const now = millis();
+  int networkStatus = 0;
+  if (now - lastConnectionTickTime > connectionTickTimeInterval) {
+    switch (netConnectionState) {
+      case CONNECTION_STATE_INIT:
+        if (ss_pin == -1) {
+          networkStatus = Ethernet.begin(mac);
+        } else {
+          networkStatus = Ethernet.begin(mac, ss_pin);
+        }
+        networkStatus = Ethernet.hardwareStatus();
+        *msgBuffer = 0;
+        sprintf(msgBuffer, "Eth hardware status(): %d", networkStatus);
+        debugMessage(msgBuffer, 2);
+        if (networkStatus == EthernetNoHardware) {
+          debugMessage("No Ethernet chip connected", 0);
+          // don't continue:
+          changeConnectionState(CONNECTION_STATE_ERROR);
+          lastConnectionTickTime = now;
+          return;
+        }
+        networkStatus = Ethernet.linkStatus();
+        *msgBuffer = 0;
+        sprintf(msgBuffer, "Eth link status(): %d", networkStatus);
+        debugMessage(msgBuffer, 2);
+        if (networkStatus == LinkOFF) {
+          debugMessage("Failed to configure Ethernet via dhcp", 0);
+          // don't continue:
+          changeConnectionState(CONNECTION_STATE_ERROR);
+          lastConnectionTickTime = now;
+          return;
+        }
+        *msgBuffer = 0;
+        sprintf(msgBuffer, "Ethernet shield recognized: ID", Ethernet.hardwareStatus());
+        debugMessage(msgBuffer, 0);
+        changeConnectionState(CONNECTION_STATE_CONNECTING);
+        break;
+      case CONNECTION_STATE_CONNECTING:
+        *msgBuffer = 0;
+        sprintf(msgBuffer, "Connecting via dhcp");
+        debugMessage(msgBuffer, 2);
+        if (ss_pin == -1) {
+          networkStatus = Ethernet.begin(mac);
+        } else {
+          networkStatus = Ethernet.begin(mac, ss_pin);
+        }
+        *msgBuffer = 0;
+        sprintf(msgBuffer, "Ethernet.status(): %d", networkStatus);
+        debugMessage(msgBuffer, 2);
+        if (networkStatus == 0) {
+          *msgBuffer = 0;
+          sprintf(msgBuffer, "Connection failed");
+          debugMessage(msgBuffer, 0);
+
+          *msgBuffer = 0;
+          sprintf(msgBuffer, "Retrying in  \"%d\" milliseconds", connectionTickTimeInterval);
+          debugMessage(msgBuffer, 2);
+          //changeConnectionState(CONNECTION_STATE_CONNECTING);
+          return;
+        } else {
+          *msgBuffer = 0;
+          sprintf(msgBuffer, "Connected!");
+          debugMessage(msgBuffer, 2);
+          changeConnectionState(CONNECTION_STATE_GETTIME);
+          return;
+        }
+        break;
+      case CONNECTION_STATE_GETTIME:
```

Comment:
Add an exit condition from the CONNECTION_STATE_GETTIME state in a similar way of the WiFiConnectionManager.